Feature/2024-01-19-yannick-horrix-opentofu. Choosing Your Orchestrator: OpenTofu vs Terraform (#850)
ordina-jworks/ordina-jworks.github.io: _posts/2024-01-19-yannick-horrix-opentofu.md

---
layout: post
authors: [yannick-horrix]
title: 'Choosing Your Orchestrator: OpenTofu vs Terraform'
image: /img/2024-01-19-yannick-horrix-opentofu/banner.jpg
tags: [ cloud, automation, terraform, opentofu, iac ]
category: cloud
comments: true
---

> > The power of Open Source is the power of the people. The people rule.

Reviewer: Nice!

{:.no_toc}
- TOC
{:toc}

> # Introduction

Reviewer: ```suggestion
# Introduction & Timeline
```

Terraform, an industry-leading infrastructure as code (IaC) tool, has been a cornerstone of cloud provisioning and infrastructure management since its inception in 2014. 
Over the years, it has fostered a vibrant community comprising thousands of users, contributors, vendors, and an extensive ecosystem of open source modules and extensions.

> However, August 10, 2023, marked a significant shift in Terraform’s trajectory when HashiCorp, its stewarding organization, changed the license from the Mozilla Public License (v2.0) to the Business Source License (v1.1). 

Reviewer: ```suggestion
August 10th 2023 markes a significant shift in Terraform’s trajectory when HashiCorp, its stewarding organization, changed the license from the Mozilla Public License (v2.0) to the Business Source License (v1.1). 
```

> This abrupt change sent ripples across the Terraform community, sparking concerns about its future.

Reviewer: Discuss impact of the license change, especially w.r.t. the uncertain future of the "open-source"-ness and Free Open Source Software-ness of the product/solution

Reviewer: Reference that this will be discussed later (just read it a few paragraphs down)

In response, the open source community initiated the OpenTofu project, which was adopted by the Linux Foundation on September 20th. OpenTofu, starting off as a fork from Terraform 1.5.6, aims to create a truly open source alternative to Terraform. 

> On January 10, 2024, OpenTofu achieved a pivotal milestone by releasing its first production-stable version, OpenTofu 1.6.0, bringing new features and promising a seamless migration for current Terraform users.

Reviewer: 1.6.0. in on parity with Terraform, feature wise. New features will only land later? https://opentofu.org/blog/opentofu-is-going-ga/#whats-next

In this article we'll delve into a comparative analysis of OpenTofu and Terraform, exploring the impact of the license change, to help you make an informed decision when choosing the right tool for your infrastructure needs.
# A Comparative Analysis: OpenTofu vs Terraform
## Features
In terms of functionality, both tools are currently equivalent. 
The OpenTofu community has aimed to maintain feature parity with Terraform for the time being. 
As such, both currently offer the same large set of features making them well suited tools for IaC workloads.

> Some small features were added with the release of OpenTofu 1.6.0 however, including:

Reviewer: These are changes over the 1.5.6. release (forked version). 
Not different from vanilla TF?

* An advanced testing feature for improved stability in configurations and modules
* Enhanced S3 state backend with new authentication methods
* A new provider and module registry, allowing a simpler method of publishing via pull request

The upcoming version 1.7 aims to introduce more community requested features not available in Terraform, including:
* Client-side state encryption for heightened security in regulated environments
* Parameterizable backends, providers, and modules to enable more readable, DRY code
* Third-party extensibility, with a plugin system for new state backends

On Terraform's side, version 1.6 implemented performance improvements to the 'terraform test' command, as well as changes to the S3 backend.

For now, both tools are functionally equivalent and compatible. 

> Time will tell whether this will remain the case as development continues on both sides.

Reviewer: Discuss that it's unlikely (nor the goal) that both project will remain (feature) compatible moving forward.

## Configuration and Syntax
The core difference between OpenTofu and Terraform lies in its syntax. 
OpenTofu uses a declarative syntax, where users can define their desired end state without having to worry about the steps required to get there. 

> Terraform, on the other hand, employs a procedural approach. 

Reviewer: ? Both use HCL as their syntax? Does this section discuss the internal implementation of the solutions?

> This means users have to define the specific steps, leading to higher control over the execution at the cost of higher complexity.

Reviewer: Not true for neither TF nor OpenTofu

> In terms of configuration, OpenTofu focuses on simplicity in its configuration language, prioritizing readability and ease of use. 

Reviewer: Again, there is no difference... They are essentially the same code base.

Terraform configuration, on the other hand, can be more verbose and complex to maintain.
## Licensing and Pricing

> Both Terraform and OpenTofu are open-source tools, providing free usage.

Reviewer: `Business Source License` (used by Terraform) is not an open-source license. It's a source-available license model https://fossa.com/blog/business-source-license-requirements-provisions-history/

Reviewer: OpenTofu uses MPL 2.0 (in release 1.6.0)

As said, OpenTofu uses the Mozilla Public License (MPL), which allows developers to freely use, modify, and distribute software for both commercial and non-commercial use.

Terraform has now shifted to the Business Source License (BUSL). This also permits free use of source code, including commercial use, except when it’s used to provide an offering that competes with HashiCorp.
An example of this could be running Terraform in a hosted way in CI/CD and offering this as a production service.

> So while, Terraform remains free to use in your projects, there can be an issue when incorporating Terraform in a service offering.

Reviewer: This isn't even fully clarified nor disclosed by Hashicorp. The license opens end-users of Terraform up to potential license costs in the future, if Hashicorp wants to. 

The license is very restrictive and allows Hashicorp to "change the rules" along the way. You might be allowed to use Terraform today by Hashicorp, but there are no guarantees that that's the case tomorrow, even without a license change.

Reviewer: E.g. using a Github Action for Terraform is very much a grey zone. Technically, only Hashicorp can create Github Actions, as creating one of your own would directly compete with Hashicorp as they [have an action available](https://github.com/marketplace/actions/hashicorp-setup-terraform).

> In terms of paid offerings, Terraform does offer multiple options with Terraform Cloud, a platform managing Terraform code, availaible as a hosted service. 

Reviewer: Make pricing a separate section. Elaborate more on the impact of the license change on end-users.

Additionally, organizations with advanced security and compliance needs can purchase Terraform Enterprise. 

> This is a self-hosted distribution of Terraform Cloud which offers a private instance that includes the advanced features available in Terraform Cloud.

Reviewer: Add pricing details, referencing a date, so people get an idea of the impact of using TF Cloud on your cloud bill (it's significant since the recent resource based pricing)

## Community and Support

> Terraform currently holds the advantage when it comes to community support. 

Reviewer: ```suggestion
Today, the Terraform holds the advantage when it comes to community support. 
```

The Terraform community is vast, offering a large breadth of resources. 

> This means that users have more readily accessible examples at their disposal and can find solutions to their problems more quickly.

Reviewer: The impact of the license changes are starting to show, but only time will tell what the final impact is on the community. 
Community contributions to Terraform have almost dried up completely https://thenewstack.io/open-source-in-numbers-the-terraform-license-change-impact-on-contribution/

Reviewer: Explain that Terraform is backed by Hashicorp as a company.

OpenTofu, on the other hand, is a growing tool with an expanding community. 
Although not as large as Terraform's, its community is rapidly expanding as more users move to OpenTofu. 

> This will provide ever increasing support and resources for the tool in the future.

Reviewer: Multiple companies have publicly pledged support to OpenTofu, including `Sopra Steria AS` !

## Maturity and Stability
Terraform is proven as a mature, stable and extremely popular orchestration tool capable of handling enterprise-grade infrastructure deployments.

> OpenTofu, as a fork of Terraform, is a newer tool with a relatively newer and may be more susceptible to bugs and instability. However, the “OpenTofu” community is actively addressing these issues, and its stability is expected to improve with time.

Reviewer: Technically, it's a very similar code base. So any instability would likely exist in both. Only for new features (like testing) this might be relevant.

Reviewer: Most of the complexity (and "moving parts) of Terraform resides in the providers anyway, which is out of scope of the license change "at time of writing"

## Summary
To summarize, the following table captures the comparison between the two tools.

| Feature | OpenTofu | Terraform |
|----------|----------|----------|
| Features | Similar to Terraform, with some improvements | Large featureset |

> | Configuration | Declarative, easier to use | Procedural, more complex |

Reviewer: Incorrect, same syntax

| Licensing | Mozilla Public License (MPL), open-source | Business Software License (BUSL), open-source with restrictions |

> | Pricing | Free to use | Free to use, with some restrictions for commercial use |

Reviewer: Terraform "Free to use, with restrictive commercial use"

> | Community | Smaller, but expanding | Large |

Reviewer: ```suggestion
| Community | Smaller, but expanding | Large, but support is shrinking due to sentiment |
```

> | Maturity | New | Mature |

Reviewer: ```suggestion
```

> | Maturity | New | Mature |

Reviewer: Add 
| Ownership | Part of Linux Foundation | Owned by Hashicorp |
| Open Source | Yes | Source-available |

# How to Migrate?
OpenTofu promises a seamless migration for Terraform users. 
As both OpenTofu 1.6 and Terraform 1.6 are compatible, the migration process is relatively straightforward. 
However, it is advised to have a disaster recovery plan in place as the migration still poses a non-trivial change to the architecture. 
The migration process is detailed in the [OpenTofu Migration Guide](https://opentofu.org/docs/intro/migration/){:target="_blank" rel="noopener noreferrer"}.

In summary, the process looks like this:
1. Apply all changes with Terraform

   `terraform apply`
2. [Install OpenTofu](https://opentofu.org/docs/install/){:target="_blank" rel="noopener noreferrer"}
   
    Test if you can successfully execute the `tofu` command:

    `tofu --version`
3. Back up your state file
   
   If you are using a local state file, you can simply make a copy of your `terraform.tfstate` file in your project directory.

>    If you are using a remote backend such as an S3 bucket, make sure that you follow the backup procedures for the backend and that you exercise the restore procedure at least once. 

Reviewer: Include `Terraform state pull` command to easily backup remote states to local machines https://developer.hashicorp.com/terraform/cli/commands/state/pull

4. Initialize OpenTofu

   `tofu init`
5. Inspect the Plan

   `tofu plan`
6. Test out a small change

   `tofu apply`

And that's it.
Migration complete!

> In order to rollback, it suffices to create a backup of the new state file and perform

Reviewer: This rollback scenario only works if you revert between the same versions (1.6.0)

`terraform init`

`terraform plan`

to verify that everything is working correctly.
# Conclusion
Despite the emergence of OpenTofu, Terraform remains a stalwart in the IaC realm. 
HashiCorp's enduring commitment to innovation, its reputation as an industry standard, and a vast user community ensure Terraform's relevance. 
While the licensing shift prompted reflection, HashiCorp continues to shape the IaC landscape, adapting to industry dynamics.

> As OpenTofu steps into the limelight, its future is laden with promises. 

Reviewer: ChatGPT wording ? ;)

The community's rapid growth, the Linux Foundation's stewardship, and the release of version 1.6 signal a trajectory of continuous improvement and innovation. 
Features like advanced testing, enhanced S3 state backend, and a new provider and module registry underscore OpenTofu's commitment to evolving alongside user needs.

> The upcoming OpenTofu 1.7 holds exciting prospects, including community-requested features not available in Terraform, such as client-side state encryption, parameterizable backends, providers, and modules, and third-party extensibility through a plugin system for new state backends.

Reviewer: This will also likely signal the end of feature compatibility between both.

> When you are a service provider looking to incorporate Terraform in one of your offerings, it can be worthwhile to look at OpenTofu as an alternative to avoid licensing costs or legal trouble.

Reviewer: And rely on a true open source product

For regular users, not much has changed on the surface. 
Terraform still is a very robust and well supported too capable of handling your orchestration needs. 
OpenTofu does offer some benefits in ease of use and configuration, and as both tools are currently compatible and functionally equivalent it may be interesting to take try out OpenTofu and see what it has to offer. 

The open-source community supporting OpenTofu will decide the future of the project. 
While the early signs are positive, it remains to be seen whether the tool can keep up in functionality and compatibility with Terraform, or even surpass it in some areas. 
At Jworks, We are certainly interested in its future development.

# What's next?

> In a follow-up post, we will attempt to migrate some of our projects from Terraform to OpenTofu. 

Reviewer: Perfect!

An example of this could be running Terraform in a hosted way in CI/CD and offering this as a production service.

So while, Terraform remains free to use in your projects, there can be an issue when incorporating Terraform in a service offering.

## Pricing and Offerings

In terms of pricing, both tools are free to use, with some limitations for commercial use when it comes to Terraform.

Hashicorp does offer paid solutions, starting with Terraform Cloud, a centralized platform for managing Terraform code. Key features include version control integration, secure variable storage, remote state management, and policy enforcement, enabling organizations to efficiently maintain control over their cloud infrastructure.

[Pricing](https://www.hashicorp.com/products/terraform/pricing?product_intent=terraform){:target="_blank" rel="noopener noreferrer"} for Terraform Cloud follows a Resources Under Management (RUM) model, where Terraform counts the amount of objects it manages and calculates cost accordingly. 
The Standard Tier offering is billed at $0.00014 per hour per resource. For the Plus and Enterprise tiers, pricing is negotiated directly with Hashicorp sales.

> Let's take the example os Terraform Enterprise. This is a self-hosted solution offered as a private installation rather than a SaaS solution.

Reviewer: ```suggestion
Let's take the example for Terraform Enterprise. 
This is a self-hosted solution offered as a private installation rather than a SaaS solution.
```

Judging from the AWS Marketplace and Azure Marketplace offerings of Terraform Enterprise, pricing starts from $15,000/year. 
This includes just five workspaces, likely to be insufficieint for a large enterprise.

## Community and Support

Terraform currently holds the advantage when it comes to community support. 

> Backed by Hashicorp and benefiting from years of community-wide support as an open-source project pre-license change), it can offer a large breadth of resources for user to take advantage of.

Reviewer: ```suggestion
Backed by Hashicorp and benefiting from years of community-wide support as an open-source project pre-license change. 
It offers a large breadth of resources for user to take advantage of.
```

OpenTofu, on the other hand, is a growing tool with an expanding community. 
Although not as large as Terraform's, its community is rapidly expanding as more users move to OpenTofu.
It is backed by the Linux Foundation as well as companies such as Gruntwork, env0, Scalr and our very own Sopra Steria AS.
This growing support will provide ever increasing resources for users of the tool in the future.

## Maturity and Stability

Terraform is proven as a mature, stable and extremely popular orchestration tool capable of handling enterprise-grade infrastructure deployments.
OpenTofu is a new tool. However, as a fork of Terraform, it stems from the same code base and thus is expected to perform similarly in terms of stability.

## Summary

To summarize, the following table captures the comparison between the two tools.

| Feature | OpenTofu | Terraform |
|----------|----------|----------|
| Features | Similar to Terraform, with some improvements | Large featureset |

> | Configuration | Declarative, easier to use | Declarative but more complex |

Reviewer: ```suggestion
| Configuration | Declarative, easier to use | Declarative but more complex |
```

Remove, it's effectively the same configuration?
